src/cascadia/TerminalApp/TerminalTab.cpp

```diff
@@ -920,87 +920,159 @@ namespace winrt::TerminalApp::implementation
     // - control: the TermControl to add events to.
     // Return Value:
     // - <none>
-    void TerminalTab::_AttachEventHandlersToControl(const uint32_t paneId, const TermControl& control)
+    void TerminalTab::_AttachEventHandlersToContent(const uint32_t paneId, const TerminalApp::IPaneContent& content)
     {
         auto weakThis{ get_weak() };
         auto dispatcher = TabViewItem().Dispatcher();
-        ControlEventTokens events{};
-
-        events.titleToken = control.TitleChanged([dispatcher, weakThis](auto&&, auto&&) -> winrt::fire_and_forget {
-            // The lambda lives in the `std::function`-style container owned by `control`. That is, when the
-            // `control` gets destroyed the lambda struct also gets destroyed. In other words, we need to
-            // copy `weakThis` onto the stack, because that's the only thing that gets captured in coroutines.
-            // See: https://devblogs.microsoft.com/oldnewthing/20211103-00/?p=105870
-            const auto weakThisCopy = weakThis;
-            co_await wil::resume_foreground(dispatcher);
-            if (auto tab{ weakThisCopy.get() })
-            {
-                // The title of the control changed, but not necessarily the title of the tab.
-                // Set the tab's text to the active panes' text.
-                tab->UpdateTitle();
-            }
-        });
+        ContentEventTokens events{};
+
+        events.CloseRequested = content.CloseRequested(
+            winrt::auto_revoke,
+            [dispatcher, weakThis](auto sender, auto&&) -> winrt::fire_and_forget {
+                // Don't forget! this ^^^^^^^^ sender can't be a reference, this is a async callback.
+
+                // The lambda lives in the `std::function`-style container owned by `control`. That is, when the
+                // `control` gets destroyed the lambda struct also gets destroyed. In other words, we need to
+                // copy `weakThis` onto the stack, because that's the only thing that gets captured in coroutines.
+                // See: https://devblogs.microsoft.com/oldnewthing/20211103-00/?p=105870
+                const auto weakThisCopy = weakThis;
+                co_await wil::resume_foreground(dispatcher);
+                // Check if Tab's lifetime has expired
+                if (auto tab{ weakThisCopy.get() })
+                {
+                    if (const auto content{ sender.try_as<TerminalApp::IPaneContent>() })
+                    {
+                        tab->_rootPane->WalkTree([content](std::shared_ptr<Pane> pane) {
+                            if (pane->GetContent() == content)
+                            {
+                                pane->Close();
+                            }
+                        });
+                    }
+                }
+            });

```

Comment:
FYI this addition breaks #16598 because it closes all tabs when it receives a `AppHost::_CloseRequested` event and then serializes them to disk. This code however asynchronously emits another `AppHost::_CloseRequested` event which causes the entire process to restart but now there are no tabs left and so it ends up with an empty persistence state.

I think this issue also exists without my persistence PR, however there it's not apparent since nothing of importance happens in `AppHost::_CloseRequested`.

Reply:
I've fixed the issue by simply not shutting down tabs on exit.

However, I believe that this is a flaw in this changeset regardless. This is because this event hook creates a loop in my opinion: When the tab closes, it closes the pane which closes the content which closes the pane which closes the tab. This recursion may end up emitting events twice, but definitely at least does things redundantly (even if the 2nd time around may not have any obvious effect).

For instance, `TerminalPage::_RemoveTab` is called when you explicitly close a tab. This will `Shutdown()` the root pane which will close the content. This will asynchronously emit a `Closed` to the pane which, if it still exists for whatever reason, bubbles it up to the `TerminalPage` which got the `Closes` event hooked up to `_RemoveTab`.